_posts/2022-03-25-ppo-implementation-details.md

```diff
@@ -171,33 +171,37 @@ We first introduce the 13 core implementation details commonly used regardless o
+    - In this architecture, PPO first initializes a **vectorized environment** `envs` that runs $N$ (usually independent) environments either sequentially or in parallel by leveraging multi-processes. `envs` presents a synchronous interface that always outputs a batch of $N$ observations from $N$ environments, and it takes a batch of $N$ actions to step the $N$ environments. When calling `next_obs = envs.reset()`, `next_obs` gets a batch of $N$ initial observations (pronounced "next observation"). PPO also initializes an environment done flag variable `next_done` (pronounced "next done") to an $N$-length array of zeros, where its i-th element `next_done[i]` has values of 0 or 1 which corresponds to the $i$-th sub-environment being *not done* and *done*, respectively.
+    - Then, the vectorized architecture loops two phases: the **rollout phase** and the **learning phase**:
+        - Rollout phase : The agent samples actions for the $N$ environments and continue to step them for a fixed number of $M$ steps. During these $M$ steps, the agent continues to append relevant data in an empty list `data`. If the $i$-th sub-environment is done (terminated or truncated) after stepping with the $i$-th action `action[i]`, `envs` would set its returned `next_done[i]` to 1, auto-reset the $i$-th sub-environment and fill `next_obs[i]` with the initial observation in the new episode of the $i$-th environment.
+        - Learning phase: The agent in principal learns from the collected data in the rollout phase: `data` of length $NM$, `next_obs` and `done`. Specifically, PPO can estimate value for the next observation `next_obs` conditioned on `next_done` and calculate the advantage `advantages` and the return `returns`, both of which also has length $NM$. PPO then learns from the prepared data `[data, advantages, returns]`, which is called "fixed-length trajectory segments" by [(Schulman et al., 2017)](#Schulman2017). 
+        - **It is important to understand `next_obs` and `next_done`'s role to help transition between phases**: At the end of the $j$-th rollout phase, `next_obs` can be used to estimate value in the learning phase, and in the begining of the $(j+1)$-th rollout phase, `next_obs` becomes the initial observation in `data`. This intricate design allows PPO to continue step the sub-environments, and because agent always learns from fixed-length trajectory segments after $M$ steps, PPO can train the agent even if the sub-environments never terminate or truncate. This is in principal why PPO can learn in long-horizon games that last 100,000 steps [(default truncation limit for Atari games in `gym`)](https://github.com/openai/gym/blob/a7b6462136ebaa610c8941e4da8a9c92155b04d1/gym/envs/__init__.py#L744) in a single episode.
```

Comment:
```suggestion
        - **It is important to understand `next_obs` and `next_done`'s role to help transition between phases**: At the end of the $j$-th rollout phase, `next_obs` can be used to estimate the value of the final state during learning phase, and in the begining of the $(j+1)$-th rollout phase, `next_obs` becomes the initial observation in `data`. This intricate design allows PPO to continue step the sub-environments, and because agent always learns from fixed-length trajectory segments after $M$ steps, PPO can train the agent even if the sub-environments never terminate or truncate. This is in principal why PPO can learn in long-horizon games that last 100,000 steps [(default truncation limit for Atari games in `gym`)](https://github.com/openai/gym/blob/a7b6462136ebaa610c8941e4da8a9c92155b04d1/gym/envs/__init__.py#L744) in a single episode.
```